src/V3Error.cpp

```diff
@@ -103,7 +103,9 @@ void V3ErrorGuarded::vlAbortOrExit() VL_REQUIRES(m_mutex) {
         std::cerr << msgPrefix() << "Aborting since under --debug" << endl;
         V3Error::vlAbort();
     } else {
-        std::exit(1);
+        // Exit without triggering any global destructors.
+        // Used to prevent detached V3ThreadPool jobs accessing destroyed static objects.
+        ::_exit(1);
```

Comment:
This will not close files, or check any assertions on destructors, and other nasties.  Maybe instead override the "delete" destructor in the thread pool object?

Reply:
The V3Error is not buffered therefore it will not be a problem here. Additionally, all log messages are flushed with `std::endl` after each line so log inconsistency is not possible. The only way that the file may remain not flushed is for the emitted files in the `V3EmitC`, however when some error occurs during their multi-threaded generation we cannot ensure this consistency. Some workers generating those files may be finished but it could happen that not all of them, this is expected since we want to stop them when some error occur to avoid waiting and emitting (probably) garbage.

Reply:
We intentionally do not always use endl, as flushes are slow.
Another problem is the coverage will not get dumped, VL_GCOV_DUMP() needs to be called.

Maybe I'm paranoid but I expect this to cause painful bugs, why can't we do the delete trick or otherwise?

Reply:
I see. This implementation does not change when coverage is flushed since the `::_exit` is called when `V3Error:debugDefault()` is false and on this path there is no `VL_GCOV_DUMP` invocation. The coverage is only flushed when `V3Error:debugDefault()` is true, thus the vlAbort is called with `std::abort`. This part remains untouched.

Reply:
Normal exit calls VL_GCOV_DUMP automatically, while _exit does not.  But one example of _exit hazards.

Reply:
As I understand it, `VL_GCOV_DUMP()` is called in the debug build, and if we're in a debug build and get an error, we abort, so it sounds like `VL_GCOV_DUMP()` isn't called anyway.

Ultimately what we need here is that errors are presented fully and that we exit with a non-zero exit code. Errors are not buffered, so that's not a problem (even if they were, they are emitted from the exiting thread, and would be flushed before exiting).

Does it matter that a C++ file is not emitted fully if we're erroring out? In a multi-threaded environment, even with flushing, it's not guaranteed that we emit full files anyway.

Using `_exit()` is a response to painful bugs like deadlocks and complex code that ultimately only served to appease C++'s RAII style, not solve a real problem.

What do you mean by the "delete trick"?

Reply:
Normal exit() calls atexit hooks, which the profiler (outside Verilator) inserts to call the function inside VL_GCOV_DUMP() automatically.  VL_GCOV_DUMP therefore needs to be called on any "exit-like" path not using exit().

>Does it matter that a C++ file is not emitted fully if we're erroring out?

Yes, this greatly helps log-based debug.

>What do you mean by the "delete trick"?
>>Some workers generating those files may be finished but it could happen that not all of them, this is expected since we want to stop them when some error occur to avoid waiting and emitting (probably) garbage.

I meant you could have an "operator delete()" or perhaps just destructor that for threads at exit time just didn't really call the threads's destructor.

BTW I don't object to calling _exit() on a thread dying in some case where we really have to skip some steps, my objection is calling _exit on every even single-threaded vlAbortOrExit.  If go this route add an argument to vlAbortOrExit to indicate the short-circuiting?

Reply:
Sure, `VL_GCOV_DUMP()` can be called manually before `_exit()`.

> Yes, this greatly helps log-based debug.

C++ files are not flushed on `std::exit()` anyway. They are owned by emit visitors which live on the stack. They don't get destroyed, which can be proven by adding an `exit()` somwehere in `V3Emit*` and placing a breakpoint in `V3OutFile::~V3OutFile()`. If we want to handle that, we need to use exceptions (in both single- and multi-threaded paths).

>  If go this route add an argument to vlAbortOrExit to indicate the short-circuiting?

Just to clarify, you're OK with calling `std::exit()` in single-threaded mode, and `_exit()` in multi-threaded? We can do that.

Reply:
>Just to clarify, you're OK with calling std::exit() in single-threaded mode, and _exit() in multi-threaded? 

If possible, std::exit in single-threaded, or in multi-threaded only inside the main-thread, and _exit otherwise.

Reply:
Note this matches what one usually has to do similarly with fork().